src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+              stream.write(buffer);
+              let httpContent = new BFSE.HttpContent({ type: media.contentType, contentLength: buffer.length }, stream);
+              httpContentList.push(httpContent);
+            }))
+        .count()
+        .flatMap(_ => {
+          let url = `/v3/directline/conversations/${this.conversationId}/users/${messageWithoutAttachments.from.id}/upload`;
+          let request = BFSE.StreamingRequest.create('PUT', url);
+          var activityStream = new BFSE.SubscribableStream();
+          activityStream.write(JSON.stringify(messageWithoutAttachments), 'utf-8');
+          request.addStream(new BFSE.HttpContent({ type: "application/vnd.microsoft.activity", contentLength: activityStream.length }, activityStream));
+          httpContentList.forEach(e => request.addStream(e));
+          return this.streamConnection.send(request);
+        })
+        .do(resp => {
+          if (resp.streams && resp.streams.length !== 1) {
+            subscriber.error("Invalid stream count " + resp.streams.length);
+          } else {
+            resp.streams[0].readAsJson()
+              .then(json => {
+                subscriber.next(json['Id'])
+              })
+          }
+        })
+        .subscribe(_ => _) // force execution
+    })
+  }
+
+  private disconnectionHandler(e: any) {
+    if (this.connectionStatus$.value === ConnectionStatus.Connecting ||
+        this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    if (this.token === null) {
+      this.activitySubscriber.error("Token unavailable");
+      return;
+    }
+
+    this.connectionStatus$.next(ConnectionStatus.Connecting);
+    this.retryCount--;
+    if (this.retryCount > 0) {
+      setTimeout(this.streamingWebSocketActivity$.bind(this), this.getRetryDelay());
+    } else {
+      console.warn("Exhausted retries");
+      this.activitySubscriber.error(e);
+    }
+  }
+
+  private streamingWebSocketActivity$(): Observable<Activity> {
+    if (this.activitySubscriber) {
+      this.theStreamHandler.setSubscriber(this.activitySubscriber);
+      this.connectAsync().then(_ => _);
+    } else {
+      return Observable.create(async (subscriber: Subscriber<Activity>) => {
+        this.activitySubscriber = subscriber;
+        this.theStreamHandler = new StreamHandler(subscriber, this.connectionStatus$);
+        await this.connectAsync();
+      });
+    }
+  }
+
+  private async waitUntilOnline() {
+    return new Promise<void>((resolve, reject) => {
+      this.connectionStatus$.subscribe((cs) => {
+        if (cs === ConnectionStatus.Online) return resolve();
+      },
+        (e) => reject(e));
+    })
+  }
+
+  private async connectAsync() {
+    let re = new RegExp('^http(s?)');
+    if (!re.test(this.domain)) throw ("Domain must begin with http or https");
+    let wsUrl = this.domain.replace(re, "ws$1") + '/conversations/connect?token=' + this.token + '&conversationId=' + this.conversationId;
```

Comment:
```suggestion
    let wsUrl = `${this.domain.replace(re, 'ws$1')}/conversations/connect?token=${this.token}&conversationId=${this.conversationId}`;
```

Also, don't use string concat for URL. Use `new URLSearchParams()`.

```js
new URLSearchParams({ token: 'a1b2c3d', conversationId: 'a1b2c3d' }).toString()
```